torch/_export/constraints.py

```diff
@@ -1,48 +1,25 @@
-from typing import Optional, Callable, Union
+from typing import Optional
 
 import torch
-from torch import SymInt, SymFloat
-from torch._dynamo import allow_in_graph
-from torch.fx.experimental.symbolic_shapes import constrain_range_int
-from torch.utils._sympy.value_ranges import ValueRangeError
-
-# `Scalar` type used in native_functions.ymal will be translated to `Union[Number, _complex]`
-# could cause type error during since `SymInt` or `SymFloat` will be used.
-# Here manually specify the type explicitly.
-sym_constrain_range: Callable[
-    [Union[int, float, SymInt, SymFloat], Optional[int], Optional[int]],
-    None,
-] = torch.sym_constrain_range  # type: ignore[assignment]
 
 
 # TODO: we want to hide this min/max stuff under some abstraction similar to
 # DynamicDim
-@allow_in_graph
 def constrain_as_value(symbol, min: Optional[int] = None, max: Optional[int] = None):
     """
-    Add min/max constraint on the intermediate symbol at tracing time
+    Add min/max constraint on the intermediate symbol at tracing time. If called in eager mode,
+    it will still check if the input value is within the specified range.
     """
-
-    if not isinstance(symbol, SymInt):
-        constrain_range_int(symbol, min=min, max=max)
-    else:
-        sym_constrain_range(symbol, min, max)
-
-    return symbol
+    torch.sym_constrain_range(symbol, min=min, max=max)
 
 
 # TODO: we want to hide this min/max stuff under some abstraction similar to
 # DynamicDim
-@allow_in_graph
-def constrain_as_size(symbol, min: int = 2, max: Optional[int] = None):
+def constrain_as_size(symbol):
     """
-    Add min/max constraint on the intermediate symbol which will be used as a size
+    Add compiler hints to the intermediate symbol which will be used as a size
+    in another tensor.
```

Comment:
Here's the docblock I would write:

This indicates that a given int is size-like, and can be used in any context where a size is expected. You will typically use this when reading out unbacked integers from Tensors, e.g., max.item() or lengths.tolist() which then need to be used as tensor constructors. Providing these assertions to PyTorch can help resolve GuardOnDataDependentSymNode errors upon export, since we cannot guard on unbacked SymInts.

This function has unusual semantics which distinguish it from constrain_as_value. Specifically, at compile-time, we will unsoundly assume that the resulting int is always `>= 2`. This makes it easier to use the unbacked int in size contexts, as we will often attempt to guard on a size being zero/one (e.g., when computing the contiguity of a tensor, or testing if broadcasting can occur), which will not work on unbacked SymInts. Assuming that the int is `>= 2` allows us to report False to these tests. Although this is technically unsound, in practice we observe that if your program works for all sizes `>= 2`, it probably works for zero and one too. At runtime, we only assert that the size-like int is `>= 0`, matching this intuition. (This is in contrast to `constrain_as_value(s0, min=2)` which would require, at runtime, that s0 is >= 2, or `constrain_as_value(s0, min=0)`, which typically isn't enough information to let you use s0 as a size to construct tensors.)